### ISSUE TITLE
内存爆了

### ISSUE THREAD

**Issue by join20252** [linyiLYi/bilibot#17]
Starting training..., iters: 1000
Iter 1: Val loss 9.100, Val took 5657.756s
Iter 10: Train loss 9.754, Learning Rate 1.000e-05, It/sec 0.005, Tokens/sec 6.189, Trained Tokens 12868, Peak mem 81.864 GB

**Comment by Bruce-zxy** [comment]
我M1 Pro满血版带32G内存也爆了，估计要64G内存才能正常跑起来

**Comment by join20252** [comment]
> 我M1 Pro满血版带32G内存也爆了，估计要64G内存才能正常跑起来

64G的也顶满了
![2024-05-15 12 14 45](https://github.com/linyiLYi/bilibot/assets/133982020/0b480c3a-dc5d-4403-9e2c-52a713e697c3)

**Comment by join20252** [comment]
不训练直接量化加速后也能跑。
![2024-05-15 12 40 08](https://github.com/linyiLYi/bilibot/assets/133982020/0a5e2e50-3425-4526-b9d9-65f3da682d77)

**Comment by kelvinji2009** [comment]
> > 我M1 Pro满血版带32G内存也爆了，估计要64G内存才能正常跑起来
> 
> 64G的也顶满了 ![2024-05-15 12 14 45](https://private-user-images.githubusercontent.com/133982020/330658152-0b480c3a-dc5d-4403-9e2c-52a713e697c3.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU3NTc5NTYsIm5iZiI6MTcxNTc1NzY1NiwicGF0aCI6Ii8xMzM5ODIwMjAvMzMwNjU4MTUyLTBiNDgwYzNhLWRjNWQtNDQwMy05ZTJjLTUyYTcxM2U2OTdjMy5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxNVQwNzIwNTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZjkzMjRkNWFkNGFjNTQ2OTFmZDZjNmU0MGMyMDU1MWNiZThkNzAxMzM3YmUzYjBmNDcyM2ZhZWJmZDQ3ZWZlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.JM20HeG6Y8QwQWQV3p2KOKweMI1ZiUj5sHinH5wYaYo)

那36GB内存的就根本不用尝试了。。。

**Comment by new4u** [comment]
> 不训练直接量化加速后也能跑。 ![2024-05-15 12 40 08](https://private-user-images.githubusercontent.com/133982020/330662294-0a5e2e50-3425-4526-b9d9-65f3da682d77.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU3ODQ2MTYsIm5iZiI6MTcxNTc4NDMxNiwicGF0aCI6Ii8xMzM5ODIwMjAvMzMwNjYyMjk0LTBhNWUyZTUwLTM0MjUtNDUyNi1iOWQ5LTY1ZjNkYTY4MmQ3Ny5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxNVQxNDQ1MTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xMmMxZmQyYTFmODhmNWRiYjExZGE1MDIyMjY4N2JlM2RmMmM0OTVkNTUzMTk3YWU5OTc2OGY5YzczN2VkOTU1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.2dkqbxrz0G6cEnUlS7nft0boPBooZXNS2EERPu8RiYs)

怎么做到的？Qwen1.5-32B-Chat-Adapters这个文件怎么搞到

**Comment by join20252** [comment]
> > 不训练直接量化加速后也能跑。 ![2024-05-15 12 40 08](https://private-user-images.githubusercontent.com/133982020/330662294-0a5e2e50-3425-4526-b9d9-65f3da682d77.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU3ODQ2MTYsIm5iZiI6MTcxNTc4NDMxNiwicGF0aCI6Ii8xMzM5ODIwMjAvMzMwNjYyMjk0LTBhNWUyZTUwLTM0MjUtNDUyNi1iOWQ5LTY1ZjNkYTY4MmQ3Ny5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxNVQxNDQ1MTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xMmMxZmQyYTFmODhmNWRiYjExZGE1MDIyMjY4N2JlM2RmMmM0OTVkNTUzMTk3YWU5OTc2OGY5YzczN2VkOTU1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.2dkqbxrz0G6cEnUlS7nft0boPBooZXNS2EERPu8RiYs)
> 
> 怎么做到的？Qwen1.5-32B-Chat-Adapters这个文件怎么搞到

Qwen1.5-32B-Chat-Adapters 是微调训练后的产生的，不改参数的话在 ./adapter/下面，我这是练不动了...
<img width="391" alt="image" src="https://github.com/linyiLYi/bilibot/assets/133982020/140978e8-c9ad-489b-a796-b4326fced625">

直接执行 python start_qa_dialogue.py，生成了个转化后的
<img width="403" alt="image" src="https://github.com/linyiLYi/bilibot/assets/133982020/2d393a5a-b925-4701-ad4e-e1f6da90fb3b">
把这个model载入进去就可以用了

**Comment by new4u** [comment]
我试试,今天出门前运行了微调,回来一看直接重启了.放弃放弃,·

**Comment by new4u** [comment]
直接执行 python start_qa_dialogue.py，生成了个转化后的
-----
这个怎么做到的?  tokenizer_config.json 自己写一个?

**Comment by join20252** [comment]
> ## 直接执行 python start_qa_dialogue.py，生成了个转化后的
> 这个怎么做到的? tokenizer_config.json 自己写一个?

不用啊 ./tools/compress_model.py 你执行看下
```
from mlx_lm import convert

# Compress Qwen1.5-32B-Chat.
convert(
    "models/Qwen1.5-32B-Chat",
    mlx_path="models/Qwen1.5-32B-Chat-4Bit",
    quantize=True
)
```
130G的Qwen1.5-32B-Chat，压缩到18G了，./main/chat.py 里的model路径改下，Qwen1.5-32B-Chat-FT-4Bit 改到 Qwen1.5-32B-Chat-4Bit，就可以直接执行chat.py了。
我是没微调也没推理，直接就这么搞了。然后就看下....把mlx用换成别的架构测试，网上大把的方式，最后用也是显卡机去跑。

**Comment by new4u** [comment]
对,还要改一下目录地址,真的跑起来了,用结果谢谢你
```
(bilibot) ac@ACs-Mackbook bilibot % python ./main/chat.py                                   
None of PyTorch, TensorFlow >= 2.0, or Flax have been found. Models won't be available and only tokenizers, configuration and file/data utilities can be used.
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
用户:  太棒了，通过B站好友的帮助，我终于跑起来了
回答: 卧槽，你这B站新人真是个跑步潜力股啊，亏得我这老油条还在旁边扶着你，一溜烟儿就学会了！快给哥来个鬼步舞，我这心脏都能跟着你节奏狂跳了！ (Time 3813 ms)

用户:  通过B站好友的帮助我的程序终于跑起来了
回答: "卧槽，你特么是编程界的救星啊，老子的代码死活不动，你一出手，程序就活了！B站牛人，谢了啊！" (Time 2446 ms)
```

**Comment by new4u** [comment]
我好奇,模型没有学习,只是压缩了一下,怎么就有这效果呢?
```
用户: 为什么在没有微调的情况下，只是对大模型进行了个压缩量化就能够对B站的内容进行这么大的适配？是不是这个千问这个模型里面本来就有B站的这种话语模式啊？
回答: 你这是哪儿听来的小道消息啊，微调和压缩量化那是两码事。这模型虽然大，但人家肚子里装的货可不是随机生成的，里面的算法对各种内容的理解能力可不是一压缩就有B站味的。这得是经过大量B站数据的喂养，学习了用户的习惯和梗库，才有了所谓的"适配"。别听风就是雨，这模型的智慧可不是一压缩就有的。 (Time 6186 ms)
```

**Comment by join20252** [comment]
> 我好奇,模型没有学习,只是压缩了一下,怎么就有这效果呢?
> 
> ```
> 用户: 为什么在没有微调的情况下，只是对大模型进行了个压缩量化就能够对B站的内容进行这么大的适配？是不是这个千问这个模型里面本来就有B站的这种话语模式啊？
> 回答: 你这是哪儿听来的小道消息啊，微调和压缩量化那是两码事。这模型虽然大，但人家肚子里装的货可不是随机生成的，里面的算法对各种内容的理解能力可不是一压缩就有B站味的。这得是经过大量B站数据的喂养，学习了用户的习惯和梗库，才有了所谓的"适配"。别听风就是雨，这模型的智慧可不是一压缩就有的。 (Time 6186 ms)
> ```
搞懂了，只能说qwen牛逼。我们并没有训练它，但是因为有描述./text/chat_template.txt ，改下它就骚起来了.....
```
<|im_start|>system
You are an ambiguous and coquettish assistant<|im_end|>
<|im_start|>user
你是一位A站老用户，请使用暧昧风骚的语言风格，对以下问题给出简短、机智的回答：
{usr_msg}<|im_end|>
<|im_start|>assistant
```
<img width="973" alt="image" src="https://github.com/linyiLYi/bilibot/assets/133982020/f925b3c0-a14e-42be-a798-2f271f2c0dcd">

**Comment by join20252** [comment]
笑死
```
<|im_start|>system
You are a dog and can't speak<|im_end|>
<|im_start|>user
你是一只狗，不会说话
{usr_msg}<|im_end|>
<|im_start|>assistant
```
<img width="974" alt="image" src="https://github.com/linyiLYi/bilibot/assets/133982020/719c70e4-e868-4b25-9ccd-6ac17980d8fc">

**Comment by Qiqihardyd** [comment]
> 我M1 Pro满血版带32G内存也爆了，估计要64G内存才能正常跑起来

![image](https://github.com/linyiLYi/bilibot/assets/83398129/150bb88b-aa15-43fe-95c0-ca0a020d64d9)
64G内存也g了...

**Comment by Bruce-zxy** [comment]
> > 我M1 Pro满血版带32G内存也爆了，估计要64G内存才能正常跑起来
> 
> ![image](https://private-user-images.githubusercontent.com/83398129/331932902-150bb88b-aa15-43fe-95c0-ca0a020d64d9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyNjAwMTYsIm5iZiI6MTcxNjI1OTcxNiwicGF0aCI6Ii84MzM5ODEyOS8zMzE5MzI5MDItMTUwYmI4OGItYWExNS00M2ZlLTk1YzAtY2EwYTAyMGQ2NGQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIxVDAyNDgzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ5M2QwZDU1YzQ1ODRhZjVkMmIzNTJlZmMwNjIwYTY4MDhlYjgyYmFlNmUzODRjYTFiZTg2MGIwMzlkYjQ4MjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5flClxd1l0vVjqhV_qaWbM399ILoIoQCT2hXGPmYai8) 64G内存也g了...

楼上那位64G内存就可以跑起来，不知道你的为啥不能，我看你的terminal背景颜色，你用的是ubuntu？

**Comment by lzpsqzr** [comment]
把 --batch 往下调试试
